### PR TITLE
Patch llvm version suffix

### PR DESCRIPTION
**Pull request by kwryankrattiger**

### CODE REVIEW

**Comment by kwryankrattiger** [comment]
@chuckatkins

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 1]
```diff
@@ -342,6 +342,10 @@ class Llvm(CMakePackage, CudaPackage):
     patch('no_cyclades.patch', when='@10:12.0.0')
     patch('no_cyclades9.patch', when='@6:9')
 
+    # Add LLVM_VERSION_SUFFIX
+    # https://reviews.llvm.org/D115818
+    patch('llvm-version-suffix-macro.patch', when='@:13.0.0')
```
Comment (kwryankrattiger) [inline var/spack/repos/builtin/packages/llvm/package.py:347]: change to `@6:`, punt putting end limit until after patch is merged and released

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 2]
```diff
@@ -65,6 +65,10 @@ class LlvmAmdgpu(CMakePackage):
     # This is already fixed in upstream but not in 4.2.0 rocm release
     patch('fix-spack-detection-4.2.0.patch', when='@4.2.0:')
 
+    # Add LLVM_VERSION_SUFFIX
+    # https://reviews.llvm.org/D115818
+    patch('llvm-version-suffix-macro.patch', when='@:4.3.2')
```
Comment (kwryankrattiger) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py:70]: apply always, whenever AMD forks from the llvm that releases with this patch add limit

**Comment by trws** [comment]
Why is this patch necessary? If Mesa needs the information specifically for spack, why is spack not providing it?

**Comment by kwryankrattiger** [comment]
> Why is this patch necessary? If Mesa needs the information specifically for spack, why is spack not providing it?

@trws This isn't just for spack, it just came up in spack due to mesa not being able to check if llvm was a pre-release. This was specifically a problem for `llvm-amdgpu` which may branch from a llvm before the official release. This type of problem will most likely be recurrent since AMD and Intel both have forks of llvm which are not required to be based on released versions of llvm.

This is currently an open patch for llvm [here](https://reviews.llvm.org/D115818) and the corresponding dependent change in mesa [here](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/14232).

Eventually patching these via spack will not be required, but doing so now will allow other issues that may arise related to the version mismatch in llvm to be handled correctly in the mean time.

**Comment by trws** [comment]
Ok, my point is that this is not in LLVM yet, it's not going to be back ported, so it's only going to work in older versions in spack right?  As such, it might be better to either:

1. check it in cmake, since it's part of the installed cmake config file since 2018 that should work back to llvm 7 everywhere, with or without spack
2. have spack send in the define in MESA based on the version of libllvm it is using
3. check the version from llvm-config, since the suffix is applied to that version automatically by the build as well

I don't understand how managing this patch in spack helps this process rather than complicates maintenance.

**Pull request closed without merging**